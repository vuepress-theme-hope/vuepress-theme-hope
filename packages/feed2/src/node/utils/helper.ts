import {
  Logger,
  isLinkHttp,
  removeEndingSlash,
  removeLeadingSlash,
} from "vuepress-shared/node";

export const FEED_GENERATOR = "vuepress-plugin-feed2";

export const logger = new Logger(FEED_GENERATOR);

export const resolveUrl = (hostname: string, base = "", path = ""): string =>
  `${
    isLinkHttp(hostname)
      ? removeEndingSlash(hostname)
      : `https://${removeEndingSlash(hostname)}`
  }${base}${removeLeadingSlash(path)}`;

export const getImageMineType = (ext = ""): string =>
  `image/${
    ext === "jpg"
      ? "jpeg"
      : ext === "svg"
      ? "svg+xml"
      : ext === "jpeg" ||
        ext === "png" ||
        ext === "bmp" ||
        ext === "gif" ||
        ext === "webp"
      ? ext
      : ""
  }`;
