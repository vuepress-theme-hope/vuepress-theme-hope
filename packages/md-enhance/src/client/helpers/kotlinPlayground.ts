import type { App } from "vue";
import { inject } from "vue";

import type { KotlinPlaygroundOptions } from "../typings/index.js";

const DEFAULT_KOTLIN_PLAYGROUND_OPTIONS: KotlinPlaygroundOptions = {};

declare const __VUEPRESS_DEV__: boolean;

let kotlinPlaygroundOptions: KotlinPlaygroundOptions =
  DEFAULT_KOTLIN_PLAYGROUND_OPTIONS;

const kotlinPlaygroundSymbol = Symbol(
  __VUEPRESS_DEV__ ? "kotlinPlayground" : "",
);

export const defineKotlinPlaygroundConfig = (
  options: KotlinPlaygroundOptions,
): void => {
  kotlinPlaygroundOptions = options;
};

export const useKotlinPlaygroundConfig = (): KotlinPlaygroundOptions =>
  inject(kotlinPlaygroundSymbol)!;

export const injectKotlinPlaygroundConfig = (app: App): void => {
  app.provide(kotlinPlaygroundSymbol, kotlinPlaygroundOptions);
};
