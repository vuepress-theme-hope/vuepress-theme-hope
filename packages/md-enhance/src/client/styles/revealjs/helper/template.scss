@mixin normalize() {
  font-weight: normal;

  p {
    line-height: 1.3;
  }

  /* Ensure certain elements are never larger than the slide itself */
  img,
  video,
  iframe {
    max-width: 95%;
    max-height: 95%;
  }

  strong,
  b {
    font-weight: bold;
  }

  em {
    font-style: italic;
  }

  ol,
  dl,
  ul {
    display: inline-block;
    margin: 0 0 0 1em;
    text-align: start;
  }

  ol {
    list-style-type: decimal;
  }

  ul ul,
  ul ol,
  ol ol,
  ol ul {
    display: block;
    margin-left: 40px;
  }

  // stylelint-disable-next-line  no-descending-specificity
  ul {
    list-style-type: disc;

    ul {
      list-style-type: square;

      // stylelint-disable selector-max-compound-selectors
      ul {
        list-style-type: circle;
      }
      // stylelint-enable selector-max-compound-selectors
    }
  }

  dt {
    font-weight: bold;
  }

  dd {
    margin-left: 40px;
  }

  blockquote {
    position: relative;

    display: block;

    width: 70%;
    padding: 5px;

    background: rgb(255 255 255 / 5%);
    box-shadow: 0 0 2px rgb(0 0 0 / 20%);

    font-style: italic;

    p {
      &:first-child,
      &:last-child {
        display: inline-block;
      }
    }
  }

  q {
    font-style: italic;
  }

  pre {
    position: relative;

    display: block;

    width: 90%;
    padding: 0;

    box-shadow: 0 5px 15px rgb(0 0 0 / 15%);

    font-size: 0.55em;
    line-height: 1.2em;
    text-align: left;
    word-wrap: break-word;
  }

  code {
    text-transform: none;
  }

  pre code {
    display: block;

    overflow: auto;

    max-height: 400px;
    padding: 5px;

    word-wrap: normal;
  }

  a {
    text-decoration: none;
    transition: color 0.15s ease;

    &:hover {
      border: none;
      text-shadow: none;
    }
  }

  table {
    margin: auto;
    border-spacing: 0;
    border-collapse: collapse;
  }

  th {
    font-weight: bold;
  }

  th,
  td {
    padding: 0.2em 0.5em;
    border-bottom: 1px solid;
    text-align: left;

    &[align="center"] {
      text-align: center;
    }

    &[align="right"] {
      text-align: right;
    }
  }

  // stylelint-disable selector-max-compound-selectors
  tbody tr:last-child {
    th,
    td {
      border-bottom: none;
    }
  }
  // stylelint-enable selector-max-compound-selectors

  sup {
    vertical-align: super;
    font-size: smaller;
  }

  sub {
    vertical-align: sub;
    font-size: smaller;
  }

  small {
    display: inline-block;
    vertical-align: top;
    font-size: 0.6em;
    line-height: 1.2em;

    * {
      vertical-align: top;
    }
  }
}

@mixin template() {
  .reveal-viewport {
    background-color: var(--r-background-color);
  }

  .reveal {
    @include normalize;
    color: var(--r-main-color);
    font-size: var(--r-main-font-size);
    font-family: var(--r-main-font);

    @media print {
      .backgrounds {
        background-color: var(--r-background-color);
      }
    }

    ::selection {
      background: var(--r-selection-background-color);
      color: var(--r-selection-color);
      text-shadow: none;
    }

    // stylelint-disable selector-max-compound-selectors
    .slides section {
      &,
      > section {
        font-weight: inherit;
        line-height: 1.3;
      }
    }
    // stylelint-enable selector-max-compound-selectors

    h1 {
      font-size: var(--r-heading1-size);
      text-shadow: var(--r-heading1-text-shadow);
    }

    h2 {
      font-size: var(--r-heading2-size);
    }

    h3 {
      font-size: var(--r-heading3-size);
    }

    h4 {
      font-size: var(--r-heading4-size);
    }

    p {
      margin: var(--r-block-margin) 0;
    }

    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin: var(--r-heading-margin);

      color: var(--r-heading-color);

      font-weight: var(--r-heading-font-weight);
      font-family: var(--r-heading-font);
      line-height: var(--r-heading-line-height);
      letter-spacing: var(--r-heading-letter-spacing);
      text-transform: var(--r-heading-text-transform);
      text-shadow: var(--r-heading-text-shadow);
      word-wrap: break-word;

      &:last-child {
        margin-bottom: 0;
      }
    }

    blockquote {
      margin: var(--r-block-margin) auto;
    }

    pre {
      margin: var(--r-block-margin) auto;
      font-family: var(--r-code-font);
    }

    code {
      font-family: var(--r-code-font);
    }

    img {
      margin: var(--r-block-margin) 0;
    }

    a {
      color: var(--r-link-color);

      &:hover {
        color: var(--r-link-color-hover);
      }
    }

    .roll span::after {
      background: var(--r-link-color-dark);
      color: #fff;
    }

    .r-frame {
      border: 4px solid var(--r-main-color);
      box-shadow: 0 0 10px rgb(0 0 0 / 15%);
    }

    a .r-frame {
      transition: all 0.15s linear;
    }

    a:hover .r-frame {
      border-color: var(--r-link-color);
      box-shadow: 0 0 20px rgb(0 0 0 / 55%);
    }

    .controls {
      color: var(--r-link-color);
    }

    .progress {
      background: rgb(0 0 0 / 20%);
      color: var(--r-link-color);
    }

    table.hljs-ln {
      display: table;
      width: 100%;

      tr {
        border: none;

        &:nth-child(2n) {
          background: inherit !important;
        }
      }

      td {
        border: none;
      }
    }
  }
}
