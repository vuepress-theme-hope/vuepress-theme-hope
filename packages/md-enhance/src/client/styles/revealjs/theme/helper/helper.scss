@use "sass:map";
@use "sass:meta";
@use "vars";

$default-variables: meta.module-variables("vars");

@function circle-gradient($outer, $inner, $type: circle) {
  @return radial-gradient(
    center,
    $type farthest-corner,
    $inner 0%,
    $outer 100%
  );
}

@mixin lightmode-text-color($color) {
  section.has-light-background {
    &,
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      color: $color;
    }
  }
}

@mixin darkmode-text-color($color) {
  section.has-dark-background {
    &,
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      color: $color;
    }
  }
}

@mixin theme($theme, $theme-variables) {
  $variables: map.deep-merge($default-variables, $theme-variables);
  $viewport-color: map.get($variables, "viewport-color");
  $background-color: map.get($variables, "background-color");
  $main-font: map.get($variables, "main-font");
  $main-font-size: map.get($variables, "main-font-size");
  $main-color: map.get($variables, "main-color");
  $block-margin: map.get($variables, "block-margin");
  $heading-margin: map.get($variables, "heading-margin");
  $heading-font: map.get($variables, "heading-font");
  $heading-color: map.get($variables, "heading-color");
  $heading-line-height: map.get($variables, "heading-line-height");
  $heading-letter-spacing: map.get($variables, "heading-letter-spacing");
  $heading-text-transform: map.get($variables, "heading-text-transform");
  $heading-text-shadow: map.get($variables, "heading-text-shadow");
  $heading-font-weight: map.get($variables, "heading-font-weight");
  $heading1-text-shadow: map.get($variables, "heading1-text-shadow");
  $heading1-size: map.get($variables, "heading1-size");
  $heading2-size: map.get($variables, "heading2-size");
  $heading3-size: map.get($variables, "heading3-size");
  $heading4-size: map.get($variables, "heading4-size");
  $code-font: map.get($variables, "code-font");
  $link-color: map.get($variables, "link-color");
  $link-color-hover: map.get($variables, "link-color-hover");
  $selection-background-color: map.get(
    $variables,
    "selection-background-color"
  );
  $selection-color: map.get($variables, "selection-color");

  .reveal[data-theme="#{$theme}"] {
    --r-background-color: #{$background-color};
    --r-main-font: #{$main-font};
    --r-main-font-size: #{$main-font-size};
    --r-main-color: #{$main-color};
    --r-block-margin: #{$block-margin};
    --r-heading-margin: #{$heading-margin};
    --r-heading-font: #{$heading-font};
    --r-heading-color: #{$heading-color};
    --r-heading-line-height: #{$heading-line-height};
    --r-heading-letter-spacing: #{$heading-letter-spacing};
    --r-heading-text-transform: #{$heading-text-transform};
    --r-heading-text-shadow: #{$heading-text-shadow};
    --r-heading-font-weight: #{$heading-font-weight};
    --r-heading1-text-shadow: #{$heading1-text-shadow};
    --r-heading1-size: #{$heading1-size};
    --r-heading2-size: #{$heading2-size};
    --r-heading3-size: #{$heading3-size};
    --r-heading4-size: #{$heading4-size};
    --r-code-font: #{$code-font};
    --r-link-color: #{$link-color};
    --r-link-color-hover: #{$link-color-hover};
    --r-selection-background-color: #{$selection-background-color};
    --r-selection-color: #{$selection-color};

    @content;

    @if $viewport-color {
      --r-viewport-color: #{$viewport-color};
    } @else {
      --r-viewport-color: #{$background-color};
    }
  }
}
