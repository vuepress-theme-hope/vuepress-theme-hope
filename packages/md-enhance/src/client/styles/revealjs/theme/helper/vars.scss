@use "sass:color";

// Base settings for all themes that can optionally be overridden by the super-theme

// Background of the presentation
$background-color: #2b2b2b;
$viewport-color: null;

// Primary/body text
$main-color: #eee;
$main-font: "Lato", sans-serif;
$main-font-size: 40px;

// Vertical spacing between blocks of text
$block-margin: 20px;

// Headings
$heading-margin: 0 0 $block-margin 0;
$heading-font: "League Gothic", impact, sans-serif;
$heading-color: #eee;
$heading-line-height: 1.2;
$heading-letter-spacing: normal;
$heading-text-transform: uppercase;
$heading-text-shadow: none;
$heading-font-weight: normal;
$heading1-text-shadow: $heading-text-shadow;
$heading1-size: 3.77em;
$heading2-size: 2.11em;
$heading3-size: 1.55em;
$heading4-size: 1em;
$code-font: monospace;

// Links and actions
$link-color: #13daec;
$link-color-hover: color.scale($link-color, $lightness: 20%);

// Text selection
$selection-background-color: #ff5e99;
$selection-color: #fff;
