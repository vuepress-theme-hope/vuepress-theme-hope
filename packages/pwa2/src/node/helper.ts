import type { ManifestOption, PWAOptions } from "../shared";

const appendBaseHelper = (base: string, link: string): string =>
  link.startsWith("/") ? link.replace(/^\//, base) : link;

export const appendBasetoManifest = (
  base: string,
  manifest: ManifestOption
): ManifestOption => {
  if (manifest.icons)
    manifest.icons = manifest.icons.map((icon) => ({
      ...icon,
      src: appendBaseHelper(base, icon.src),
    }));

  if (manifest.shortcuts)
    manifest.shortcuts = manifest.shortcuts.map((shortcut) => ({
      ...shortcut,
      icons:
        shortcut.icons?.map((icon) => ({
          ...icon,
          src: appendBaseHelper(base, icon.src),
        })) || [],
      url: appendBaseHelper(base, shortcut.url),
    }));

  if (manifest.screenshots)
    manifest.screenshots = manifest.screenshots.map((screenshot) => ({
      ...screenshot,
      src: appendBaseHelper(base, screenshot.src),
    }));

  return manifest;
};

export const appendBase = (base: string, options: PWAOptions): void => {
  if (options.favicon)
    options.favicon = appendBaseHelper(base, options.favicon);

  if (options.apple) {
    if (options.apple.icon)
      options.apple.icon = appendBaseHelper(base, options.apple.icon);

    if (options.apple.maskIcon)
      options.apple.maskIcon = appendBaseHelper(base, options.apple.maskIcon);
  }

  if (options.msTile && options.msTile.image)
    options.msTile.image = appendBaseHelper(base, options.msTile.image);

  if (options.manifest)
    options.manifest = appendBasetoManifest(base, options.manifest);
};
