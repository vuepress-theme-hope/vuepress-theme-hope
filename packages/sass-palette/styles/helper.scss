@use "sass:color";
@use "sass:list";
@use "sass:map";
@use "sass:meta";
@use "sass:string";

@function is-upper-case($char) {
  @return string.to-lower-case($char) != $char;
}

@function is-lower-case($char) {
  @return string.to-upper-case($char) != $char;
}

@function camel-to-kebab($string) {
  $length: string.length($string);
  $property: "--";

  @for $i from 1 through $length {
    $char: string.slice($string, $i, $i);

    $property: "#{$property}#{if(is-upper-case($char), "-#{string.to-lower-case(
        $char
      )}",$char)}";
  }

  @return $property;
}

@function color-is-legal($value) {
  // single color
  @if meta.type-of($value) == color {
    @return true;
  }

  // itâ€™s a valid map color config
  @if (meta.type-of($value) == map) {
    @if (meta.type-of(map.get($value, "light")) == color) and
      (meta.type-of(map.get($value, "dark")) == color)
    {
      @return true;
    }

    @if (is-legal-config(map.get($value, "base"))) {
      $scale: map.get($value, "scale");

      @if (meta.type-of($scale) == map) {
        @each $name, $amount in $scale {
          @if (meta.type-of($amount) != number) {
            @debug "Unexpected value: getting #{$value}";

            @return false;
          }
        }

        @return true;
      }
    }
  }

  @warn "Unexpected value: getting #{$value}";

  @return false;
}

@function get-color($color, $isDark: false) {
  @if meta.type-of($color) == color {
    @return $color;
  }

  $lightColor: map.get($color, "light");
  $darkColor: map.get($color, "dark");

  @if $lightColor and $darkColor {
    @return if($isDark, $darkColor, $lightColor);
  }

  @warn "Unexpected color: getting #{$color}";

  @return transparent;
}

@mixin inject-color($key, $config, $dark-selector: "") {
  @if meta.type-of($config) == color {
    :root {
      #{$key}: #{$config};
    }
  } @else {
    $lightColor: map.get($config, "light");
    $darkColor: map.get($config, "dark");

    @if $lightColor and $darkColor {
      :root {
        #{$key}: #{$lightColor};
      }

      @if $lightColor != $darkColor and $dark-selector {
        #{$dark-selector} {
          #{$key}: #{$darkColor};
        }
      }
    }
  }
}
