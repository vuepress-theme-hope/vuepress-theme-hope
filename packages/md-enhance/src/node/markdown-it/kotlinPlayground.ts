import { encodeData, entries } from "@vuepress/helper";
import type { PluginSimple } from "markdown-it";

import { playground } from "./playground/index.js";
import type { PlaygroundData } from "../typings/index.js";

export const kotlinPlayground: PluginSimple = (md) => {
  md.use(playground, {
    name: "kotlin-playground",
    component: "KotlinPlayground",
    propsGetter: ({ title = "", key, files, settings }: PlaygroundData) => ({
      title,
      key,
      settings: encodeURIComponent(JSON.stringify(settings || {})),
      files: encodeData(
        JSON.stringify(entries(files).map(([, { content }]) => content)),
      ),
    }),
  });
};
