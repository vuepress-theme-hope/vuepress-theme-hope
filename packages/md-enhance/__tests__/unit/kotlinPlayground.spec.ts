/* eslint-disable @typescript-eslint/naming-convention */
import { decodeData } from "@vuepress/helper";
import MarkdownIt from "markdown-it";
import { describe, expect, it } from "vitest";

import { kotlinPlayground } from "../../src/node/markdown-it/kotlinPlayground.js";

const decodeFiles = (content: string): string[] =>
  JSON.parse(decodeData(content)) as string[];

const getFiles = (renderResult: string): string[] | null => {
  const result = renderResult.match(/files="(.*?)"/s);

  if (!result) return null;

  return decodeFiles(result[1]);
};

const getSettings = (renderResult: string): Record<string, unknown> | null => {
  const result = renderResult.match(/settings="(.*?)"/s);

  if (!result) return null;

  return JSON.parse(decodeURIComponent(result[1])) as Record<string, unknown>;
};

describe("Kotlin Playground", () => {
  const markdownIt = MarkdownIt({ linkify: true }).use(kotlinPlayground);

  it("Should resolve playground info", () => {
    const result = markdownIt.render(
      `
::: kotlin-playground Playground title

@file main.kt

\`\`\`kotlin
class Contact(val id: Int, var email: String)

fun main(args: Array<String>) {
    val contact = Contact(1, "mary@gmail.com")
    println(contact.id)
}
\`\`\`

:::

`,
      {},
    );

    expect(result).toMatchSnapshot();

    const files = getFiles(result);
    const settings = getSettings(result);

    expect(files).toEqual([
      `\
class Contact(val id: Int, var email: String)

fun main(args: Array<String>) {
    val contact = Contact(1, "mary@gmail.com")
    println(contact.id)
}
`,
    ]);
    expect(settings).toEqual({});
  });

  it("Should resolve playground info with settings", () => {
    const result = markdownIt.render(
      `
::: kotlin-playground Readonly Playground

@file main.kt

\`\`\`kotlin
import cat.Cat

fun main(args: Array<String>) {
//sampleStart
    val cat = Cat("Kitty")
    println(cat.name)
//sampleEnd
}
\`\`\`

@file cat.kt

\`\`\`kotlin
package cat
class Cat(val name: String)
\`\`\`


@settings

\`\`\`json
{
  "data-highlight-only": ""
}
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();

    const file = getFiles(result);
    const settings = getSettings(result);

    expect(file).toEqual([
      `\
import cat.Cat

fun main(args: Array<String>) {
//sampleStart
    val cat = Cat("Kitty")
    println(cat.name)
//sampleEnd
}
`,
      `\
package cat
class Cat(val name: String)
`,
    ]);
    expect(settings).toEqual({
      "data-highlight-only": "",
    });
  });
});
