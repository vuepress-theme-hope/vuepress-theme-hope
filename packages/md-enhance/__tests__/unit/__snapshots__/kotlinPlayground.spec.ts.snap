// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Kotlin Playground > Should resolve playground info 1`] = `
"<KotlinPlayground title="Playground%20title" key="16b0e07e" settings="%7B%7D" files="eJw9zDELwjAUBOC/cmRKoRRcg4ri5OxoHB6phkDyKi+xUMT/bkLFG++476pcpJxxmriQK3qmiDAanLn0mElwTxSiwaVIYN9Ztvx4MWrJmsRng6MILdt133d4W0ZNc9xqYvfXNz2sSiTLwTd2cFOyqqLt8axAiax/ryGMdfhYVrcvD2o3zg==">
</KotlinPlayground>
"
`;

exports[`Kotlin Playground > Should resolve playground info with settings 1`] = `
"<KotlinPlayground title="Readonly%20Playground" key="2012bfab" settings="%7B%22data-highlight-only%22%3A%22%22%7D" files="eJw1zrEKAjEMgOFXCZ3uQLz9UEHEyfFG4xBqLcU2ljYKh/juph52bPi/5GxCyo8iYEnWBxJk5NuTIVHgjoqvI+xLoXkzSQnsdz28kYehUsrRTUJFC9D3otgI2IIiHZpTEJnR9Ms0ayuRu7aEKTn9/htHviJ/kM3KZLJ38q45yDZSrT+s0S0aYblBY3P5AgEPP2Y=">
</KotlinPlayground>
"
`;
